File: agents/path-planning-agent/backend/app/api/routes/optimizer.py
"""模块四：路径规划 API 路由"""
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional, Dict, Any, List
from app.services.optimizer.optimizer_service import run_optimizer
from app.services.optimizer.algorithms.aco.solver import get_default_params
from app.services.optimizer.models.task_model import parse_task
from app.services.optimizer.models.solution import Solution
from app.services.optimizer.engine.distance_matrix import build_distance_matrix
from app.services.optimizer.algorithms.aco.colony import solve_with_aco
from app.services.optimizer.outputs.excel_exporter import export_excel
from app.core.database import SessionLocal
from app.models.optimizer import OptimizationRecord
import json
from datetime import datetime

router = APIRouter(prefix="/api/optimizer", tags=["模块四-路径规划"])


class OptimizeRequest(BaseModel):
    task: Dict[str, Any]
    aco_params: Optional[Dict[str, Any]] = None


class DefaultParamsRequest(BaseModel):
    task: Dict[str, Any]


@router.post("/run")
def run_optimize(req: OptimizeRequest):
    """
    运行路径优化。

    请求体:
        task: 任务 JSON（包含 depot、demand_points、distance_matrix、uavs）
        aco_params: ACO 参数（可选，不传则使用默认智能参数）

    返回:
        优化结果（summary、route_table、village_table、drone_table、geojson）
    """
    if not req.task.get("demand_points"):
        raise HTTPException(status_code=400, detail="任务必须包含需求点")
    if not req.task.get("uavs"):
        raise HTTPException(status_code=400, detail="任务必须包含至少一架无人机")

    try:
        result = run_optimizer(req.task, req.aco_params)
        
        # 保存优化记录到数据库
        try:
            db = SessionLocal()
            summary = result.get("summary", {})
            record = OptimizationRecord(
                task_config=json.dumps(req.task, ensure_ascii=False),
                solution_data=json.dumps(result, ensure_ascii=False),
                aco_params=json.dumps(req.aco_params, ensure_ascii=False) if req.aco_params else None,
                summary=json.dumps(summary, ensure_ascii=False),
                total_distance=summary.get("total_distance", 0),
                total_energy=summary.get("total_energy", 0),
                total_trips=summary.get("total_trips", 0),
                village_count=summary.get("village_count", 0),
                drone_count=summary.get("drone_count", 0),
            )
            db.add(record)
            db.commit()
            db.refresh(record)
            db.close()
        except Exception as e:
            print(f"[Optimizer] 保存记录失败: {e}")
        
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"路径优化失败: {type(e).__name__}: {str(e)}")


@router.post("/params/default")
def get_default(req: DefaultParamsRequest):
    """获取默认 ACO 参数"""
    try:
        task = parse_task(req.task)
        params = get_default_params(task)
        return params
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/export/excel")
def export_excel_endpoint(req: OptimizeRequest):
    """
    导出 Excel 运输汇总（4个Sheet）。

    返回: Excel 文件流
    """
    if not req.task.get("demand_points"):
        raise HTTPException(status_code=400, detail="任务必须包含需求点")

    try:
        task = parse_task(req.task)
        aco_params = req.aco_params or get_default_params(task)
        distance_matrix = build_distance_matrix(task)
        solution = solve_with_aco(task, aco_params, distance_matrix)
        excel_bytes = export_excel(task, solution)

        return StreamingResponse(
            iter([excel_bytes]),
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={"Content-Disposition": "attachment; filename=transport_plan.xlsx"},
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"导出失败: {type(e).__name__}: {str(e)}")


# ─── 历史记录 CRUD ───

@router.get("/history")
def get_optimization_history(limit: int = 20):
    """
    获取路径规划历史记录
    
    参数:
        limit: 返回记录数量（默认20条）
    
    返回:
        优化记录列表
    """
    try:
        db = SessionLocal()
        records = db.query(OptimizationRecord).order_by(
            OptimizationRecord.created_at.desc()
        ).limit(limit).all()
        
        result = []
        for record in records:
            result.append({
                "id": record.id,
                "total_distance": record.total_distance,
                "total_energy": record.total_energy,
                "total_trips": record.total_trips,
                "village_count": record.village_count,
                "drone_count": record.drone_count,
                "created_at": record.created_at.isoformat() if record.created_at else None,
            })
        
        db.close()
        return {"records": result}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取历史记录失败: {str(e)}")


@router.get("/{record_id}")
def get_optimization_detail(record_id: int):
    """
    获取单个优化记录详情
    
    参数:
        record_id: 记录ID
    
    返回:
        优化记录详情（包含完整方案数据）
    """
    try:
        db = SessionLocal()
        record = db.query(OptimizationRecord).filter(OptimizationRecord.id == record_id).first()
        
        if not record:
            db.close()
            raise HTTPException(status_code=404, detail="记录不存在")
        
        result = {
            "id": record.id,
            "task_config": json.loads(record.task_config) if record.task_config else None,
            "solution_data": json.loads(record.solution_data) if record.solution_data else None,
            "aco_params": json.loads(record.aco_params) if record.aco_params else None,
            "summary": json.loads(record.summary) if record.summary else None,
            "total_distance": record.total_distance,
            "total_energy": record.total_energy,
            "total_trips": record.total_trips,
            "village_count": record.village_count,
            "drone_count": record.drone_count,
            "created_at": record.created_at.isoformat() if record.created_at else None,
        }
        
        db.close()
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取记录详情失败: {str(e)}")


@router.delete("/{record_id}")
def delete_optimization(record_id: int):
    """
    删除优化记录
    
    参数:
        record_id: 记录ID
    
    返回:
        删除结果
    """
    try:
        db = SessionLocal()
        record = db.query(OptimizationRecord).filter(OptimizationRecord.id == record_id).first()
        
        if not record:
            db.close()
            raise HTTPException(status_code=404, detail="记录不存在")
        
        db.delete(record)
        db.commit()
        db.close()
        
        return {"success": True, "message": "记录已删除"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"删除记录失败: {str(e)}")

File: frontend/src/stores/pathPlanning/optimizer.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import {
  runOptimizer as apiRun,
  getDefaultParams as apiDefaults,
  exportExcel as apiExportExcel,
  getOptimizationHistory as apiHistory,
  getOptimizationDetail as apiDetail,
  deleteOptimization as apiDelete,
  manualAdjust as apiManual,
} from '@/api/pathPlanning/optimizer'
import { usePointsStore } from './points'
import { useMaterialsStore } from './materials'
import { useUavsStore } from './uavs'

const PRIORITY_MAP = { 1: 'urgent', 2: 'high', 3: 'medium', 4: 'low', 5: 'normal' }

export const useOptimizerStore = defineStore('optimizer', () => {
  // ─── State ───
  const loading = ref(false)
  const error = ref(null)
  const result = ref(null)
  const acoParams = ref(null)
  const defaultParams = ref(null)
  const history = ref([])
  const historyLoading = ref(false)
  const selectedDroneId = ref(null) // 当前选中查看路线的无人机ID
  const regenHints = ref(null)      // 合规核验回填的提示（{hints:[], params:{}}），Module4 展示后清除
  const originalResult = ref(null)  // ACO 算法原始结果快照（手动调序后可回退）
  const manualActive = ref(false)   // 当前结果是否为手动调序版本

  // ─── Getters ───
  const summary = computed(() => result.value?.summary || {})
  const routeTable = computed(() => result.value?.route_table || [])
  const villageTable = computed(() => result.value?.village_table || [])
  const droneTable = computed(() => result.value?.drone_table || [])
  const geojson = computed(() => result.value?.geojson || null)
  const elapsed = computed(() => result.value?.elapsed_seconds || 0)

  // 兼容旧 getter
  const routes = computed(() => result.value?.solution?.trips || [])
  const totalDistance = computed(() => summary.value?.total_distance || 0)
  const totalTime = computed(() => summary.value?.total_time || 0)
  const totalTrips = computed(() => summary.value?.total_trips || 0)
  const droneCount = computed(() => summary.value?.drone_count || 0)
  const villageCount = computed(() => summary.value?.village_count || 0)
  const allCovered = computed(() => summary.value?.feasible || false)

  // 可行性校验
  const feasibility = computed(() => result.value?.feasibility || null)
  const feasible = computed(() => feasibility.value?.feasible ?? false)
  const feasibleText = computed(() => feasibility.value?.feasible_text || '')
  const feasibilityIssues = computed(() => feasibility.value?.issues || [])
  const feasibilityWarnings = computed(() => feasibility.value?.warnings || [])
  const demandCoverage = computed(() => feasibility.value?.demand_coverage || {})
  const tripChecks = computed(() => feasibility.value?.trip_checks || {})

  // ─── Actions ───

  /** 从 stores 构建 task JSON */
  function buildTaskJson() {
    const ptsStore = usePointsStore()
    const matStore = useMaterialsStore()
    const uavStore = useUavsStore()

    if (!ptsStore.center) throw new Error('请先设置配送中心')
    if (ptsStore.demands.length === 0) throw new Error('请先添加需求点')
    if (uavStore.selections.length === 0) throw new Error('请先选择无人机')

    const depot = {
      id: ptsStore.center.id || 'depot',
      name: ptsStore.center.name,
      longitude: ptsStore.center.longitude,
      latitude: ptsStore.center.latitude,
    }

    const demand_points = ptsStore.demands.map((pt) => {
      const assignment = matStore.getAssignment(pt.id)
      const materials = []
      if (assignment?.items) {
        for (const item of assignment.items) {
          const w = item.subtotal ?? (item.unit_weight ?? item.weight ?? 0) * (item.qty ?? item.quantity ?? 1)
          materials.push({
            type: item.name || item.category || '物资',
            weight: w || 0,
          })
        }
      }
      if (materials.length === 0 && assignment?.total_weight) {
        materials.push({ type: '物资', weight: assignment.total_weight })
      }
      return {
        id: pt.id,
        name: pt.name,
        longitude: pt.longitude,
        latitude: pt.latitude,
        priority: PRIORITY_MAP[assignment?.priority] || 'medium',
        delivery_mode: assignment?.delivery_mode || 'optional',
        materials,
      }
    })

    const distance_matrix = ptsStore.distMatrix.length > 0 ? ptsStore.distMatrix : []

    const uavs = []
    for (const sel of uavStore.selectedDetails) {
      const m = sel.model
      console.log('[Optimizer] UAV Model:', m.id, m.model, m.range_points)
      for (let i = 0; i < sel.quantity; i++) {
        uavs.push({
          id: `${m.id}-${i + 1}`,
          name: m.model || m.name,
          max_payload: m.max_payload,
          max_range: m.range_km || m.max_range,
          battery_capacity: 100,
          max_speed: m.max_speed || 60,
          range_points: m.range_points || [],
        })
      }
    }
    console.log('[Optimizer] Task JSON UAVs:', uavs)

    return { depot, demand_points, distance_matrix, uavs }
  }

  /** 获取默认参数 */
  async function fetchDefaultParams() {
    try {
      const task = buildTaskJson()
      const res = await apiDefaults(task)
      defaultParams.value = res.data
      acoParams.value = { ...res.data }
      return res.data
    } catch (e) {
      error.value = e.message
      return null
    }
  }

  /** 运行优化 */
  async function runOptimization(customParams = null) {
    loading.value = true
    error.value = null
    result.value = null
    try {
      const task = buildTaskJson()
      const params = customParams || acoParams.value
      const res = await apiRun(task, params)
      result.value = res.data
      // 保存算法原始结果快照，供手动调序后回退
      originalResult.value = res.data
      manualActive.value = false
      return res.data
    } catch (e) {
      error.value = e.response?.data?.detail || e.message
      return null
    } finally {
      loading.value = false
    }
  }

  /**
   * 手动调整某航次的配送顺序并重算
   * @param {number} tripId 目标航次 id（solution.trips 中的 trip_id）
   * @param {number[]} newRoute 重排后的节点序列（0 开头结尾）
   */
  async function adjustTripOrder(tripId, newRoute) {
    const trips = (result.value?.solution?.trips || []).map((t) =>
      t.trip_id === tripId ? { ...t, route: newRoute } : t
    )
    loading.value = true
    error.value = null
    try {
      const task = buildTaskJson()
      const res = await apiManual(task, trips)
      result.value = res.data
      manualActive.value = true
      return res.data
    } catch (e) {
      error.value = e.response?.data?.detail || e.message
      return null
    } finally {
      loading.value = false
    }
  }

  /** 回退到 ACO 算法原始结果 */
  function revertToAlgorithm() {
    if (originalResult.value) {
      result.value = originalResult.value
      manualActive.value = false
    }
  }

  /** 导出 Excel */
  async function downloadExcel() {
    try {
      const task = buildTaskJson()
      const params = acoParams.value
      const res = await apiExportExcel(task, params)
      const blob = new Blob([res.data], {
        type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
      })
      const url = URL.createObjectURL(blob)
      const a = document.createElement('a')
      a.href = url
      a.download = `运输方案_${Date.now()}.xlsx`
      a.click()
      URL.revokeObjectURL(url)
    } catch (e) {
      error.value = e.message
    }
  }

  /** 重置结果 */
  function resetResult() {
    result.value = null
    error.value = null
    originalResult.value = null
    manualActive.value = false
  }

  /** 加载历史记录 */
  async function loadHistory(limit = 20) {
    historyLoading.value = true
    try {
      const res = await apiHistory(limit)
      history.value = res.data.records || []
      return history.value
    } catch (e) {
      console.error('加载优化历史失败:', e)
      return []
    } finally {
      historyLoading.value = false
    }
  }

  /** 查看历史记录详情 */
  async function viewHistoryDetail(recordId) {
    try {
      const res = await apiDetail(recordId)
      const data = res.data
      if (data.solution_data) {
        result.value = data.solution_data
      }
      return data
    } catch (e) {
      error.value = e.response?.data?.detail || e.message
      return null
    }
  }

  /** 删除历史记录 */
  async function deleteRecord(recordId) {
    try {
      await apiDelete(recordId)
      // 从本地列表中移除
      history.value = history.value.filter(r => r.id !== recordId)
      return true
    } catch (e) {
      error.value = e.response?.data?.detail || e.message
      return false
    }
  }

  /** 设置选中的无人机ID（用于地图高亮） */
  function setSelectedDroneId(droneId) {
    selectedDroneId.value = droneId
  }

  /** 清除核验回填提示 */
  function clearRegenHints() {
    regenHints.value = null
  }

  return {
    loading,
    error,
    result,
    acoParams,
    defaultParams,
    history,
    historyLoading,
    selectedDroneId,
    regenHints,
    manualActive,
    summary,
    routeTable,
    villageTable,
    droneTable,
    geojson,
    elapsed,
    routes,
    totalDistance,
    totalTime,
    totalTrips,
    droneCount,
    villageCount,
    allCovered,
    feasibility,
    feasible,
    feasibleText,
    feasibilityIssues,
    feasibilityWarnings,
    demandCoverage,
    tripChecks,
    buildTaskJson,
    fetchDefaultParams,
    runOptimization,
    adjustTripOrder,
    revertToAlgorithm,
    downloadExcel,
    resetResult,
    loadHistory,
    viewHistoryDetail,
    deleteRecord,
    setSelectedDroneId,
    clearRegenHints,
  }
})
